### PR TITLE
Nomenclature: clearly distinguish between resolve and fulfill

### LINKED ISSUE #334: Issue with "Promise.resolve()": …
MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve

The entire description was difficult to understand.

You need to realise that your audience may have a traditional software engineering background, based on clear logical explanations. You may want to re-write your description so that it is more logical and readable, and not so self-referential.

For beginners to JavaScript Promises, more reference to other fundamental information about Promises would be useful. Some of the terminology (e.g. "flattens nested layers") needs to be properly explained, because it is not within the usual vernacular of software engineering.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/javascript/reference/global_objects/promise/resolve`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve
* GitHub URL: https://github.com/mdn/content/blob/master/files/en-us/web/javascript/reference/global_objects/promise/resolve/index.html
* Report started: 2020-12-17T01:01:36.175Z

</details>

### LINKED ISSUE #17009: Uses "resolved case" instead of "fulfilled case".
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

### What specific section or headline is this issue about?

Chained Promises

### What information was incorrect, unhelpful, or incomplete?

The `.then()` method takes up to two arguments; the first argument is a callback function for the **resolved** case of the promise, and the second argument is a callback function for the rejected case.

### What did you expect to see?

The `.then()` method takes up to two arguments; the first argument is a callback function for the **fulfilled** case of the promise, and the second argument is a callback function for the rejected case.

### Do you have any supporting links, references, or citations?

`Promise.prototype.then ( onFulfilled, onRejected )` §27.2.5.4 of the ECMA_262 Standard 12th edition.

### Do you have anything more you want to share?

_No response_

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/javascript/reference/global_objects/promise`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global_objects/promise/index.md
* Last commit: https://github.com/mdn/content/commit/9ff0ec4eb02f1629a200ce90b73a7d5151403583
* Document last modified: 2022-05-27T02:51:44.000Z

</details>

### PR DESCRIPTION
**Pull request by Josh-Cena**
#### Motivation

Fix #334 
Fix #17009

#### Metadata

- [ ] Adds a new document
- [ ] Rewrites (or significantly expands) a document
- [x] Fixes a typo, bug, or other error

### CODE REVIEW

File: files/en-us/web/javascript/reference/global_objects/promise/index.md [hunk 1]
```diff
@@ -107,18 +115,39 @@ const promiseC = promiseA.then(handleFulfilled2, handleRejected2);
...
 // produces output in this order:
 // immediate logging
 // asynchronous logging has val: 777
 ```
 
+### Thenables
+
+The JavaScript ecosystem has made multiple Promise implementations long before it became part of the language. Despite being represented differently internally, at the minimum, all Promise-like objects implement the _Thenable_ interface. A thenable implements the [`.then()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then) method, which is called with two callbacks: one for when the promise is fulfilled, one for when it's rejected. Promises are thenables as well.
+
+To interoperate with the existing Promise implementations, the language allows using thenables in place of promises. For example, [`Promise.resolve`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve) will not only resolve promises, but also trace thenables.
```
Comment (sideshowbarker) [inline files/en-us/web/javascript/reference/global_objects/promise/index.md:134]:
```suggestion
The JavaScript ecosystem had made multiple Promise implementations long before it became part of the language. Despite being represented differently internally, at the minimum, all Promise-like objects implement the _Thenable_ interface. A thenable implements the [`.then()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then) method, which is called with two callbacks: one for when the promise is fulfilled, one for when it's rejected. Promises are thenables as well.

To interoperate with the existing Promise implementations, the language allows using thenables in place of promises. For example, [`Promise.resolve`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve) will not only resolve promises, but also trace thenables.
```

**Review by sideshowbarker — APPROVED** [review]

**Comment by hamishwillee** [comment]
@sideshowbarker Ah, I've just been following the existing pattern of referring to resolved and rejected for the settled states of a non-pending promise. But sounds like the options are actually fulfilled, rejected and pending.

Resolved, according to the spec is "A promise is resolved if it is settled or if it has been “locked in” to match the state of another promise."
So I get the settled case - it means "fulfilled or rejected". What does "locked in to match the state of another promise." mean?
"

**Comment by Josh-Cena** [comment]
- "Resolved" is a "fate", not a "state"
- "unresolved" implies "pending", "settled" implies "resolved"
- When you do `const p = Promise.resolve(x)`, the `p` is an "actual promise", and `p`'s eventual state will match that of `x`. In this case we say `x` is "resolved" because you get hold of a promise describing the eventual completion of `x`, but `x` may not have completed yet.
- In practice "resolved" is AFAICT not very useful since it's mostly done behind the scenes by the `resolve` callback of `new Promise` or by the `Promise#then` method. We would almost always want to say "a promise is fulfilled", since that's when the `then` handler gets invoked.

**Comment by hamishwillee** [comment]
Thanks @Josh-Cena.  I guess MDN summarises this somewhere? 
I understand the term resolved now, but as you say it isn't very useful.

**Comment by Josh-Cena** [comment]
It's _somehow_ described here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#description (Edited in this PR) That "resolved" paragraph is mainly pointing to the [States and Fates](https://github.com/domenic/promises-unwrapping/blob/master/docs/states-and-fates.md) document, but frankly it's quite abstract. I don't really like linking to stack overflow answers on MDN but I don't know better resources for explaining this either, apart from the Promise/A+ spec.

**Comment by hamishwillee** [comment]
Thanks @Josh-Cena. FWIW I'm not sure inking to the states and fates doc adds any value in that doc. I also find the phrasing around  "Its state at this instant is also called its eventual state, ...". At that instant it is the actual state - before that instant it is the eventual state :-).

**Comment by Josh-Cena** [comment]
I agree "states and fates" is worth replacing with a better resource, if one exists (or we can write our own description if any one of us is committed enough—I'm not sure if I can explain it well either, so I lazily kept that source...)

> At that instant it is the actual state

It means "the state after it becomes not pending"—"at that instant", "after the transition from pending", et al. sound the same, but yeah I struggled with an exact definition there. If you have better wordings I'd love to hear.

**Comment by hamishwillee** [comment]
> A pending promise can either be fulfilled with a value or rejected with a reason (error).

To 

> The _eventual state_ of a pending promise can either be fulfilled with a value or rejected with a reason (error). ...

and remove "Its state at this instant is also called its eventual state, which can either be "fulfilled with some value" or "rejected with some reason"."

**Comment by Josh-Cena** [comment]
Ah, sounds good to me.

**Comment by hamishwillee** [comment]
OK, I'll do that ^^^

**Pull request merged**